File: KSConverter/Converter.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports KS.TextWriterColor
Imports KS.ColorTools
Imports KS.Kernel
Imports KS.KernelTools
Imports KS.RemoteDebugTools
Imports KS.NetworkTools
Imports KS.AliasManager
Imports KS.UserManagement
Imports KS.PermissionManagement
Imports KS.Config
Imports System.IO
Imports FluentFTP

Module Converter

    ''' <summary>
    ''' Main entry point
    ''' </summary>
    Sub Main()
        'Check for terminal (macOS only). This check is needed because we have the stock Terminal.app (Apple_Terminal according to $TERM_PROGRAM) that
        'has incompatibilities with VT sequences, causing broken display. It claims it supports XTerm, yet it isn't fully XTerm-compliant, so we exit
        'the program early when this stock terminal is spotted.
#If STOCKTERMINALMACOS = False Then
        If IsOnMacOS() Then
            If GetTerminalEmulator() = "Apple_Terminal" Then
                Console.WriteLine("Kernel Simulator makes use of VT escape sequences, but Terminal.app has broken support for 255 and true colors. This program can't continue.")
                Environment.Exit(5)
            End If
        End If
#End If

        'Initialize all needed variables
        Dim ListOfOldPaths = GetOldPaths("")
        Dim ListOfBackups = GetOldPaths("KSBackup")

        'Initialize paths
        InitPaths()

        'Load user token
        LoadUserToken()

        'Make backup directory
        Write("- Making backup directory...", True, ColTypes.Stage)
        If Not Directory.Exists(GetHomeDirectory() + "/KSBackup") Then
            'Just make it!
            Write("  - Backup directory not found. Creating directory...", True, ColTypes.Neutral)
            Directory.CreateDirectory(GetHomeDirectory() + "/KSBackup")
        Else
            'Directory found. Skip the creation.
            Write("  - Warning: backup directory is already found.", True, ColTypes.Warning)
        End If
        Console.WriteLine()

        'Make backup of old configuration files in case something goes wrong during conversion.
        Write("- Making backup of old configuration files...", True, ColTypes.Stage)
        For Each ConfigEntry As String In ListOfOldPaths.Keys
            If File.Exists(ListOfOldPaths(ConfigEntry)) And Not File.Exists(ListOfBackups(ConfigEntry)) Then
                'Move the old config file to backup
                Write("  - {0}: {1} -> {2}", True, ColTypes.Neutral, ConfigEntry, ListOfOldPaths(ConfigEntry), ListOfBackups(ConfigEntry))
                File.Move(ListOfOldPaths(ConfigEntry), ListOfBackups(ConfigEntry))
            ElseIf File.Exists(ListOfBackups(ConfigEntry)) Then
                Write("  - Warning: {0} already exists", True, ColTypes.Warning, ListOfBackups(ConfigEntry))
            Else
                'File not found. Skip it.
                Write("  - Warning: {0} not found in home directory.", True, ColTypes.Warning, ListOfOldPaths(ConfigEntry))
            End If
        Next
        Console.WriteLine()

        'Import all blocked devices to DebugDeviceNames.json
        Write("- Importing all blocked devices to DebugDeviceNames.json...", True, ColTypes.Stage)
        If File.Exists(ListOfBackups("BlockedDevices")) Then
            'Read blocked devices from old file
            Write("  - Reading blocked devices from blocked_devices.csv...", True, ColTypes.Neutral)
            Dim BlockedDevices As List(Of String) = File.ReadAllLines(ListOfBackups("BlockedDevices")).ToList
            Write("  - {0} devices found.", True, ColTypes.Neutral, BlockedDevices.Count)

            'Add blocked devices to new format
            For Each BlockedDevice As String In BlockedDevices
                Write("  - Adding {0} to DebugDeviceNames.json...", True, ColTypes.Neutral, BlockedDevice)
                AddDeviceToJson(BlockedDevice, False)
                SetDeviceProperty(BlockedDevice, DeviceProperty.Blocked, True)
            Next
        Else
            'File not found. Skip stage.
            Write("  - Warning: blocked_devices.csv not found in home directory.", True, ColTypes.Warning)
        End If
        Console.WriteLine()

        'Import all FTP speed dial settings to JSON
        Write("- Importing all FTP speed dial addresses to FTP_SpeedDial.json...", True, ColTypes.Stage)
        If File.Exists(ListOfBackups("FTPSpeedDial")) Then
            'Read FTP speed dial addresses from old file
            Write("  - Reading FTP speed dial addresses from ftp_speeddial.csv...", True, ColTypes.Neutral)
            Dim SpeedDialLines As String() = File.ReadAllLines(ListOfBackups("FTPSpeedDial"))
            Write("  - {0} addresses found.", True, ColTypes.Neutral, SpeedDialLines.Count)

            'Add addresses to new format
            For Each SpeedDialLine As String In SpeedDialLines
                Dim ChosenLineSeparation As String() = SpeedDialLine.Split(",")
                Dim Address As String = ChosenLineSeparation(0)
                Dim Port As String = ChosenLineSeparation(1)
                Dim Username As String = ChosenLineSeparation(2)
                Dim Encryption As FtpEncryptionMode = [Enum].Parse(GetType(FtpEncryptionMode), ChosenLineSeparation(3))
                Write("  - Adding {0} to FTP_SpeedDial.json...", True, ColTypes.Neutral, Address)
                AddEntryToSpeedDial(Address, Port, Username, SpeedDialType.FTP, Encryption)
            Next
        Else
            'File not found. Skip stage.
            Write("  - Warning: ftp_speeddial.csv not found in home directory.", True, ColTypes.Warning)
        End If
        Console.WriteLine()

        'Import all users to JSON
        Write("- Importing all users to Users.json...", True, ColTypes.Stage)
        If File.Exists(ListOfBackups("Users")) Then
            'Read all users from old file
            Write("  - Reading users from users.csv...", True, ColTypes.Neutral)
            Dim UsersLines As String() = File.ReadAllLines(ListOfBackups("Users"))
            Write("  - {0} users found.", True, ColTypes.Neutral, UsersLines.Count)

            'Add users to new format
            For Each UsersLine As String In UsersLines
                Write("  - Adding {0} to Users.json...", True, ColTypes.Neutral, UsersLine.Split(",")(0))
                InitializeUser(UsersLine.Split(",")(0), UsersLine.Split(",")(1), False)
                If UsersLine.Split(",")(2) = "True" Then
                    AddPermission(PermissionType.Administrator, UsersLine.Split(",")(0))
                End If
                If UsersLine.Split(",")(3) = "True" Then
                    AddPermission(PermissionType.Disabled, UsersLine.Split(",")(0))
                End If
                If UsersLine.Split(",")(4) = "True" Then
                    AddPermission(PermissionType.Anonymous, UsersLine.Split(",")(0))
                End If
            Next
        Else
            'File not found. Skip stage.
            Write("  - Warning: users.csv not found in home directory.", True, ColTypes.Warning)
        End If
        Console.WriteLine()

        'Import all aliases to JSON
        Write("- Importing all aliases to Aliases.json...", True, ColTypes.Stage)
        If File.Exists(ListOfBackups("Aliases")) Then
            'Read all aliases from old file
            Write("  - Reading users from aliases.csv...", True, ColTypes.Neutral)
            Dim AliasesLines As String() = File.ReadAllLines(ListOfBackups("Aliases"))
            Write("  - {0} aliases found.", True, ColTypes.Neutral, AliasesLines.Count)

            'Add aliases to new format
            For Each AliasLine As String In AliasesLines
                Dim AliasLineSplit() As String = AliasLine.Split({", "}, StringSplitOptions.RemoveEmptyEntries)
                Dim AliasCommand As String = AliasLineSplit(1)
                Dim ActualCommand As String = AliasLineSplit(2)
                Dim AliasType As String = AliasLineSplit(0)
                Write("  - Adding {0} to Aliases.json...", True, ColTypes.Neutral, AliasCommand)
                Select Case AliasType
                    Case "Shell"
                        If Not Aliases.ContainsKey(AliasCommand) Then
                            Aliases.Add(AliasCommand, ActualCommand)
                        End If
                    Case "Remote"
                        If Not RemoteDebugAliases.ContainsKey(AliasCommand) Then
                            RemoteDebugAliases.Add(AliasCommand, ActualCommand)
                        End If
                    Case "FTPShell"
                        If Not FTPShellAliases.ContainsKey(AliasCommand) Then
                            FTPShellAliases.Add(AliasCommand, ActualCommand)
                        End If
                    Case "SFTPShell"
                        If Not SFTPShellAliases.ContainsKey(AliasCommand) Then
                            SFTPShellAliases.Add(AliasCommand, ActualCommand)
                        End If
                    Case "Mail"
                        If Not MailShellAliases.ContainsKey(AliasCommand) Then
                            MailShellAliases.Add(AliasCommand, ActualCommand)
                        End If
                    Case Else
                        Write("  - Invalid type {0}", True, ColTypes.Error, AliasType)
                End Select
            Next

            'Save the changes
            Write("  - Saving aliases to Aliases.json...", True, ColTypes.Neutral)
            SaveAliases()
        Else
            'File not found. Skip stage.
            Write("  - Warning: aliases.csv not found in home directory.", True, ColTypes.Warning)
        End If
        Console.WriteLine()

        'Import all config to JSON
        Write("- Importing all kernel config to KernelConfig.json...", True, ColTypes.Stage)
        If File.Exists(ListOfBackups("Configuration")) Then
            'Read all config from old file
            Write("  - Reading config from kernelConfig.ini...", True, ColTypes.Neutral)
            If Not ReadPreFivePointFiveConfig(ListOfBackups("Configuration")) Then
                If Not ReadFivePointFiveConfig(ListOfBackups("Configuration")) Then
                    Write("  - Warning: kernelConfig.ini has incompatible format. Generating new config anyways...", True, ColTypes.Warning)
                End If
            End If

            'Save the changes
            Write("  - Saving configuration to KernelConfig.json...", True, ColTypes.Neutral)
            CreateConfig()
        Else
            'File not found. Skip stage.
            Write("  - Warning: kernelConfig.ini not found in home directory.", True, ColTypes.Warning)
        End If
        Console.WriteLine()

        'Print this message:
        Write("- Successfully converted all settings to new format! Enjoy!", True, ColTypes.Stage)
        Write("- Press any key to exit.", True, ColTypes.Stage)
        Console.ReadKey(True)

    End Sub

    ''' <summary>
    ''' Gets all paths that are used in 0.0.15.x or earlier kernels.
    ''' </summary>
    ''' <param name="AppendedPath">Optional path to append</param>
    Function GetOldPaths(ByVal AppendedPath As String) As Dictionary(Of String, String)
        'Initialize all needed variables
        Dim OldPaths As New Dictionary(Of String, String)

        'Check to see if we're appending new path name
        If Not String.IsNullOrEmpty(AppendedPath) Then AppendedPath = $"/{AppendedPath}"

        'Populate our dictionary with old paths
        If IsOnUnix() Then
            OldPaths.Add("Configuration", Environ("HOME") + $"{AppendedPath}/kernelConfig.ini")
            OldPaths.Add("Aliases", Environ("HOME") + $"{AppendedPath}/aliases.csv")
            OldPaths.Add("Users", Environ("HOME") + $"{AppendedPath}/users.csv")
            OldPaths.Add("FTPSpeedDial", Environ("HOME") + $"{AppendedPath}/ftp_speeddial.csv")
            OldPaths.Add("BlockedDevices", Environ("HOME") + $"{AppendedPath}/blocked_devices.csv")
        Else
            OldPaths.Add("Configuration", Environ("USERPROFILE").Replace("\", "/") + $"{AppendedPath}/kernelConfig.ini")
            OldPaths.Add("Aliases", Environ("USERPROFILE").Replace("\", "/") + $"{AppendedPath}/aliases.csv")
            OldPaths.Add("Users", Environ("USERPROFILE").Replace("\", "/") + $"{AppendedPath}/users.csv")
            OldPaths.Add("FTPSpeedDial", Environ("USERPROFILE").Replace("\", "/") + $"{AppendedPath}/ftp_speeddial.csv")
            OldPaths.Add("BlockedDevices", Environ("USERPROFILE").Replace("\", "/") + $"{AppendedPath}/blocked_devices.csv")
        End If

        'Return it.
        Return OldPaths
    End Function

    ''' <summary>
    ''' Gets all paths that are used in 0.0.16.x or later kernels.
    ''' </summary>
    Function GetNewPaths() As Dictionary(Of String, String)
        'Initialize all needed variables
        Dim NewPaths As New Dictionary(Of String, String)

        'Populate our dictionary with old paths
        If IsOnUnix() Then
            NewPaths.Add("Configuration", Environ("HOME") + "/KernelConfig.json")
            NewPaths.Add("Aliases", Environ("HOME") + "/Aliases.json")
            NewPaths.Add("Users", Environ("HOME") + "/Users.json")
            NewPaths.Add("FTPSpeedDial", Environ("HOME") + "/FTP_SpeedDial.json")
            NewPaths.Add("DebugDevNames", Environ("USERPROFILE").Replace("\", "/") + "/DebugDeviceNames.json")
        Else
            NewPaths.Add("Configuration", Environ("USERPROFILE").Replace("\", "/") + "/KernelConfig.json")
            NewPaths.Add("Aliases", Environ("USERPROFILE").Replace("\", "/") + "/Aliases.json")
            NewPaths.Add("Users", Environ("USERPROFILE").Replace("\", "/") + "/Users.json")
            NewPaths.Add("FTPSpeedDial", Environ("USERPROFILE").Replace("\", "/") + "/FTP_SpeedDial.json")
            NewPaths.Add("DebugDevNames", Environ("USERPROFILE").Replace("\", "/") + "/DebugDeviceNames.json")
        End If

        'Return it.
        Return NewPaths
    End Function

    ''' <summary>
    ''' Gets home directory depending on platform
    ''' </summary>
    Function GetHomeDirectory() As String
        If IsOnUnix() Then
            Return Environ("HOME")
        Else
            Return Environ("USERPROFILE")
        End If
    End Function

End Module

File: KSConverter/Converter.cs

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using FluentFTP;
using KS.ConsoleBase;
using KS.ConsoleBase.Colors;
using KS.ConsoleBase.Inputs;
using KS.Files;
using KS.Files.Querying;
using KS.Login;
using KS.Misc.Configuration;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.FancyWriters;
using KS.Network;
using KS.Network.RemoteDebug;
using KS.Shell.ShellBase.Aliases;
using KS.Shell.ShellBase.Shells;
using static KS.ConsoleBase.Colors.KernelColorTools;

namespace KSConverter
{
    static class Converter
    {
        /// <summary>
        /// Main entry point
        /// </summary>
        public static void Main()
        {
            // Check console
            ConsoleSanityChecker.CheckConsole();

            try
            {
                // Initialize all needed variables
                var ListOfOldPaths = ConverterTools.GetOldPaths("");
                var ListOfBackups = ConverterTools.GetOldPaths("KSBackup");

                // Initialize paths
                Paths.InitPaths();

                // Load user token
                UserManagement.LoadUserToken();

                // Make backup directory
                SeparatorWriterColor.WriteSeparator("[1/7] Making backup directory", true, ColTypes.Stage);
                Debug.WriteLine($"Backup directory: {ConverterTools.GetHomeDirectory() + "/KSBackup"}");
                Debug.WriteLine($"FolderExists = {Checking.FolderExists(ConverterTools.GetHomeDirectory() + "/KSBackup")}");
                if (!Checking.FolderExists(ConverterTools.GetHomeDirectory() + "/KSBackup"))
                {
                    // Just make it!
                    TextWriterColor.Write("  - Backup directory not found. Creating directory...", true, GetConsoleColor(ColTypes.Progress));
                    Debug.WriteLine("Creating directory...");
                    Directory.CreateDirectory(ConverterTools.GetHomeDirectory() + "/KSBackup");
                }
                else
                {
                    // Directory found. Skip the creation.
                    Debug.WriteLine("Already found.");
                    TextWriterColor.Write("  - Warning: backup directory is already found.", true, GetConsoleColor(ColTypes.Warning));
                }
                TextWriterColor.WritePlain("", true);

                // Make backup of old configuration files in case something goes wrong during conversion.
                SeparatorWriterColor.WriteSeparator("[2/7] Making backup of old configuration files", true, ColTypes.Stage);
                foreach (string ConfigEntry in ListOfOldPaths.Keys)
                {
                    Debug.WriteLine($"Old path config entry: {ConfigEntry}");
                    Debug.WriteLine($"Old path exists: {Checking.FileExists(ListOfOldPaths[ConfigEntry])}");
                    Debug.WriteLine($"Backup exists: {Checking.FileExists(ListOfBackups[ConfigEntry])}");
                    if (Checking.FileExists(ListOfOldPaths[ConfigEntry]) & !Checking.FileExists(ListOfBackups[ConfigEntry]))
                    {
                        // Move the old config file to backup
                        Debug.WriteLine($"Moving {ConfigEntry} from {ListOfOldPaths[ConfigEntry]} to {ListOfBackups[ConfigEntry]}...");
                        TextWriterColor.Write("  - {0}: {1} -> {2}", true, GetConsoleColor(ColTypes.Neutral), ConfigEntry, ListOfOldPaths[ConfigEntry], ListOfBackups[ConfigEntry]);
                        File.Move(ListOfOldPaths[ConfigEntry], ListOfBackups[ConfigEntry]);
                    }
                    else if (Checking.FileExists(ListOfBackups[ConfigEntry]))
                    {
                        Debug.WriteLine("We already have backup!");
                        TextWriterColor.Write("  - Warning: {0} already exists", true, GetConsoleColor(ColTypes.Warning), ListOfBackups[ConfigEntry]);
                    }
                    else
                    {
                        // File not found. Skip it.
                        Debug.WriteLine("We don't have config.");
                        TextWriterColor.Write("  - Warning: {0} not found in home directory.", true, GetConsoleColor(ColTypes.Warning), ListOfOldPaths[ConfigEntry]);
                    }
                }
                TextWriterColor.WritePlain("", true);

                // Import all blocked devices to DebugDeviceNames.json
                SeparatorWriterColor.WriteSeparator("[3/7] Importing all blocked devices to DebugDeviceNames.json", true, ColTypes.Stage);
                Debug.WriteLine($"Blocked device backup exists = {Checking.FileExists(ListOfBackups["BlockedDevices"])}");
                if (Checking.FileExists(ListOfBackups["BlockedDevices"]))
                {
                    // Read blocked devices from old file
                    TextWriterColor.Write("  - Reading blocked devices from blocked_devices.csv...", true, GetConsoleColor(ColTypes.Progress));
                    Debug.WriteLine($"Calling File.ReadAllLines on {ListOfBackups["BlockedDevices"]}...");
                    var BlockedDevices = File.ReadAllLines(ListOfBackups["BlockedDevices"]).ToList();
                    Debug.WriteLine($"We have {BlockedDevices.Count} devices.");
                    TextWriterColor.Write("  - {0} devices found.", true, GetConsoleColor(ColTypes.Neutral), BlockedDevices.Count);

                    // Add blocked devices to new format
                    Debug.WriteLine($"Iterating {BlockedDevices.Count} blocked devices...");
                    foreach (string BlockedDevice in BlockedDevices)
                    {
                        Debug.WriteLine($"Adding blocked device {BlockedDevice} to the new config format...");
                        TextWriterColor.Write("  - Adding {0} to DebugDeviceNames.json...", true, GetConsoleColor(ColTypes.Progress), BlockedDevice);
                        RemoteDebugTools.AddDeviceToJson(BlockedDevice, false);
                        RemoteDebugTools.SetDeviceProperty(BlockedDevice, RemoteDebugTools.DeviceProperty.Blocked, true);
                    }
                }
                else
                {
                    // File not found. Skip stage.
                    Debug.WriteLine("We don't have file.");
                    TextWriterColor.Write("  - Warning: blocked_devices.csv not found in home directory.", true, GetConsoleColor(ColTypes.Warning));
                }
                TextWriterColor.WritePlain("", true);

                // Import all FTP speed dial settings to JSON
                SeparatorWriterColor.WriteSeparator("[4/7] Importing all FTP speed dial addresses to FTP_SpeedDial.json", true, ColTypes.Stage);
                Debug.WriteLine($"Speed dial addresses exists = {Checking.FileExists(ListOfBackups["FTPSpeedDial"])}");
                if (Checking.FileExists(ListOfBackups["FTPSpeedDial"]))
                {
                    // Read FTP speed dial addresses from old file
                    TextWriterColor.Write("  - Reading FTP speed dial addresses from ftp_speeddial.csv...", true, GetConsoleColor(ColTypes.Progress));
                    Debug.WriteLine($"Calling File.ReadAllLines on {ListOfBackups["FTPSpeedDial"]}...");
                    var SpeedDialLines = File.ReadAllLines(ListOfBackups["FTPSpeedDial"]);
                    Debug.WriteLine($"We have {SpeedDialLines.Length} addresses.");
                    TextWriterColor.Write("  - {0} addresses found.", true, GetConsoleColor(ColTypes.Neutral), SpeedDialLines.Length);

                    // Add addresses to new format
                    foreach (string SpeedDialLine in SpeedDialLines)
                    {
                        // Populate variables
                        var ChosenLineSeparation = SpeedDialLine.Split(',');
                        Debug.WriteLine($"Separation count = {ChosenLineSeparation.Length}");
                        string Address = ChosenLineSeparation[0];
                        Debug.WriteLine($"Address = {Address}");
                        string Port = ChosenLineSeparation[1];
                        Debug.WriteLine($"Port = {Port}");
                        string Username = ChosenLineSeparation[2];
                        Debug.WriteLine($"Username = {Username}");
                        FtpEncryptionMode Encryption = (FtpEncryptionMode)Enum.Parse(typeof(FtpEncryptionMode), ChosenLineSeparation[3]);
                        Debug.WriteLine($"Encryption = {Encryption}");

                        // Add the entry!
                        TextWriterColor.Write("  - Adding {0} to FTP_SpeedDial.json...", true, GetConsoleColor(ColTypes.Progress), Address);
                        NetworkTools.AddEntryToSpeedDial(Address, Convert.ToInt32(Port), Username, NetworkTools.SpeedDialType.FTP, Encryption);
                    }
                }
                else
                {
                    // File not found. Skip stage.
                    Debug.WriteLine("We don't have file.");
                    TextWriterColor.Write("  - Warning: ftp_speeddial.csv not found in home directory.", true, GetConsoleColor(ColTypes.Warning));
                }
                TextWriterColor.WritePlain("", true);

                // Import all users to JSON
                SeparatorWriterColor.WriteSeparator("[5/7] Importing all users to Users.json", true, ColTypes.Stage);
                Debug.WriteLine($"Users file exists = {Checking.FileExists(ListOfBackups["Users"])}");
                if (Checking.FileExists(ListOfBackups["Users"]))
                {
                    // Read all users from old file
                    TextWriterColor.Write("  - Reading users from users.csv...", true, GetConsoleColor(ColTypes.Progress));
                    Debug.WriteLine($"Calling File.ReadAllLines on {ListOfBackups["Users"]}...");
                    var UsersLines = File.ReadAllLines(ListOfBackups["Users"]);
                    Debug.WriteLine($"We have {UsersLines.Length} addresses.");
                    TextWriterColor.Write("  - {0} users found.", true, GetConsoleColor(ColTypes.Neutral), UsersLines.Length);

                    // Add users to new format
                    foreach (string UsersLine in UsersLines)
                    {
                        // Populate variables
                        Debug.WriteLine($"Parsing line {UsersLine}...");
                        var ChosenLineSeparation = UsersLine.Split(',');
                        Debug.WriteLine($"Separation count = {ChosenLineSeparation.Length}");
                        string Username = ChosenLineSeparation[0];
                        Debug.WriteLine($"Username = {Username}");
                        string Password = ChosenLineSeparation[1];
                        Debug.WriteLine($"Password = {Password}");
                        string Administrator = ChosenLineSeparation.Length >= 3 ? ChosenLineSeparation[2] : "False";
                        Debug.WriteLine($"Administrator = {Administrator}");
                        string Disabled = ChosenLineSeparation.Length >= 4 ? ChosenLineSeparation[3] : "False";
                        Debug.WriteLine($"Disabled = {Disabled}");
                        string Anonymous = ChosenLineSeparation.Length >= 5 ? ChosenLineSeparation[4] : "False";
                        Debug.WriteLine($"Anonymous = {Anonymous}");

                        // Add the entry!
                        TextWriterColor.Write("  - Adding {0} to Users.json...", true, GetConsoleColor(ColTypes.Progress), Username);
                        UserManagement.InitializeUser(Username, Password, false);
                        if (Administrator == "True")
                        {
                            Debug.WriteLine($"Adding the Administrator permission to {Username}...");
                            PermissionManagement.AddPermission(PermissionManagement.PermissionType.Administrator, Username);
                        }
                        if (Disabled == "True")
                        {
                            Debug.WriteLine($"Adding the Disabled permission to {Username}...");
                            PermissionManagement.AddPermission(PermissionManagement.PermissionType.Disabled, Username);
                        }
                        if (Anonymous == "True")
                        {
                            Debug.WriteLine($"Adding the Anonymous permission to {Username}...");
                            PermissionManagement.AddPermission(PermissionManagement.PermissionType.Anonymous, Username);
                        }
                    }
                }
                else
                {
                    // File not found. Skip stage.
                    Debug.WriteLine("We don't have file.");
                    TextWriterColor.Write("  - Warning: users.csv not found in home directory.", true, GetConsoleColor(ColTypes.Warning));
                }
                TextWriterColor.WritePlain("", true);

                // Import all aliases to JSON
                SeparatorWriterColor.WriteSeparator("[6/7] Importing all aliases to Aliases.json", true, ColTypes.Stage);
                Debug.WriteLine($"Aliases file exists = {Checking.FileExists(ListOfBackups["Aliases"])}");
                if (Checking.FileExists(ListOfBackups["Aliases"]))
                {
                    // Read all aliases from old file
                    TextWriterColor.Write("  - Reading users from aliases.csv...", true, GetConsoleColor(ColTypes.Progress));
                    Debug.WriteLine($"Calling File.ReadAllLines on {ListOfBackups["Aliases"]}...");
                    var AliasesLines = File.ReadAllLines(ListOfBackups["Aliases"]);
                    Debug.WriteLine($"We have {AliasesLines.Length} aliases.");
                    TextWriterColor.Write("  - {0} aliases found.", true, GetConsoleColor(ColTypes.Neutral), AliasesLines.Length);

                    // Add aliases to new format
                    foreach (string AliasLine in AliasesLines)
                    {
                        // POpulate variables
                        var AliasLineSplit = AliasLine.Split(new[] { ", " }, StringSplitOptions.RemoveEmptyEntries);
                        Debug.WriteLine($"Separation count = {AliasLineSplit.Length}");
                        string AliasCommand = AliasLineSplit[1];
                        Debug.WriteLine($"AliasCommand = {AliasCommand}");
                        string ActualCommand = AliasLineSplit[2];
                        Debug.WriteLine($"ActualCommand = {ActualCommand}");
                        string AliasType = AliasLineSplit[0];
                        Debug.WriteLine($"AliasType = {AliasType}");

                        // Add the entry!
                        TextWriterColor.Write("  - Adding {0} to Aliases.json...", true, GetConsoleColor(ColTypes.Progress), AliasCommand);
                        switch (AliasType ?? "")
                        {
                            case "Shell":
                                {
                                    Dictionary<string, string> aliases = AliasManager.GetAliasesListFromType(ShellType.Shell);
                                    if (!aliases.ContainsKey(AliasCommand))
                                        AliasManager.AddAlias(AliasCommand, ActualCommand, ShellType.Shell);
                                    break;
                                }
                            case "Remote":
                                {
                                    Dictionary<string, string> aliases = AliasManager.GetAliasesListFromType(ShellType.RemoteDebugShell);
                                    if (!aliases.ContainsKey(AliasCommand))
                                        AliasManager.AddAlias(AliasCommand, ActualCommand, ShellType.RemoteDebugShell);
                                    break;
                                }
                            case "FTPShell":
                                {
                                    Dictionary<string, string> aliases = AliasManager.GetAliasesListFromType(ShellType.FTPShell);
                                    if (!aliases.ContainsKey(AliasCommand))
                                        AliasManager.AddAlias(AliasCommand, ActualCommand, ShellType.FTPShell);
                                    break;
                                }
                            case "SFTPShell":
                                {
                                    Dictionary<string, string> aliases = AliasManager.GetAliasesListFromType(ShellType.SFTPShell);
                                    if (!aliases.ContainsKey(AliasCommand))
                                        AliasManager.AddAlias(AliasCommand, ActualCommand, ShellType.SFTPShell);
                                    break;
                                }
                            case "Mail":
                                {
                                    Dictionary<string, string> aliases = AliasManager.GetAliasesListFromType(ShellType.MailShell);
                                    if (!aliases.ContainsKey(AliasCommand))
                                        AliasManager.AddAlias(AliasCommand, ActualCommand, ShellType.MailShell);
                                    break;
                                }
                            default:
                                {
                                    TextWriterColor.Write("  - Invalid type {0}", true, GetConsoleColor(ColTypes.Error), AliasType);
                                    break;
                                }
                        }
                    }

                    // Save the changes
                    Debug.WriteLine("Saving...");
                    TextWriterColor.Write("  - Saving aliases to Aliases.json...", true, GetConsoleColor(ColTypes.Progress));
                    AliasManager.SaveAliases();
                }
                else
                {
                    // File not found. Skip stage.
                    Debug.WriteLine("We don't have file.");
                    TextWriterColor.Write("  - Warning: aliases.csv not found in home directory.", true, GetConsoleColor(ColTypes.Warning));
                }
                TextWriterColor.WritePlain("", true);

                // Import all config to JSON
                SeparatorWriterColor.WriteSeparator("[7/7] Importing all kernel config to KernelConfig.json", true, ColTypes.Stage);
                Debug.WriteLine($"Config file exists = {Checking.FileExists(ListOfBackups["Configuration"])}");
                if (Checking.FileExists(ListOfBackups["Configuration"]))
                {
#if !NETCOREAPP
                    // Read all config from old file
                    TextWriterColor.Write("  - Reading config from kernelConfig.ini...", true, KernelColorTools.ColTypes.Progress);
                    Debug.WriteLine("Reading configuration...");
                    if (!PreFivePointFive.ReadPreFivePointFiveConfig(ListOfBackups["Configuration"]))
                    {
                        if (!FivePointFive.ReadFivePointFiveConfig(ListOfBackups["Configuration"]))
                        {
                            Debug.WriteLine("Incompatible format. Both ReadPreFivePointFiveConfig and ReadFivePointFiveConfig returned False. Regenerating...");
                            TextWriterColor.Write("  - Warning: kernelConfig.ini has incompatible format. Generating new config anyways...", true, KernelColorTools.ColTypes.Warning);
                        }
                    }

                    // Save the changes
                    Debug.WriteLine("Saving...");
                    TextWriterColor.Write("  - Saving configuration to KernelConfig.json...", true, KernelColorTools.ColTypes.Progress);
                    Config.CreateConfig();
#else
                    // We need to use .NET Framework version of KSConverter to be able to fully use MadMilkman.Ini
                    // as we used it back when Kernel Simulator and MadMilkman.Ini were so tied to .NET Framework.
                    Debug.WriteLine("Config conversion needs to be done in .NET Framework.");
                    TextWriterColor.Write("  - Warning: To convert kernel configuration from kernelConfig.ini to KernelConfig.json, you need to run the .NET Framework version of KSConverter.", true, GetConsoleColor(ColTypes.Warning));
#endif
                }
                else
                {
                    // File not found. Skip stage.
                    Debug.WriteLine("We don't have file.");
                    TextWriterColor.Write("  - Warning: kernelConfig.ini not found in home directory.", true, GetConsoleColor(ColTypes.Warning));
                }
                TextWriterColor.WritePlain("", true);

                // Print this message:
                TextWriterColor.Write("- Successfully converted all settings to new format! Enjoy!", true, GetConsoleColor(ColTypes.Success));
                TextWriterColor.Write("- Press any key to exit.", true, GetConsoleColor(ColTypes.Success));
                Input.DetectKeypress();
            }
            catch (Exception ex)
            {
                TextWriterColor.Write("- Error converting settings: {0}", true, GetConsoleColor(ColTypes.Error), ex.Message);
                TextWriterColor.Write("- Press any key to exit. Stack trace below:", true, GetConsoleColor(ColTypes.Error));
                TextWriterColor.Write(ex.StackTrace, true, GetConsoleColor(ColTypes.Neutral));
				Input.DetectKeypress();
            }
        }

    }
}
